crates/engine_bibtex/src/c_api/log.rs

```diff
@@ -0,0 +1,715 @@
+use crate::c_api::auxi::{cur_aux, cur_aux_ln};
+use crate::c_api::bibs::{bib_line_num, cur_bib};
+use crate::c_api::buffer::{with_buffers, with_buffers_mut, BufTy};
+use crate::c_api::char_info::LexClass;
+use crate::c_api::cite::with_cites;
+use crate::c_api::exec::{bst_ex_warn_print, bst_ln_num_print, ExecCtx};
+use crate::c_api::hash::{with_hash, FnClass};
+use crate::c_api::history::{mark_error, mark_fatal, mark_warning};
+use crate::c_api::other::with_other;
+use crate::c_api::peekable::input_ln;
+use crate::c_api::pool::with_pool;
+use crate::c_api::scan::ScanRes;
+use crate::c_api::{
+    ttstub_output_close, ttstub_output_open, ttstub_output_open_stdout, ASCIICode, CResult,
+    FieldLoc, GlblCtx, HashPointer, StrNumber,
+};
+use std::cell::Cell;
+use std::ffi::CStr;
+use std::io::Write;
+use std::slice;
+use tectonic_io_base::OutputHandle;
+
+pub trait AsBytes {
+    fn as_bytes(&self) -> &[u8];
+}
+
+impl AsBytes for str {
+    fn as_bytes(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+impl AsBytes for String {
+    fn as_bytes(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+impl AsBytes for [u8] {
+    fn as_bytes(&self) -> &[u8] {
+        self
+    }
+}
+
+thread_local! {
+    static STANDARD_OUTPUT: Cell<Option<&'static mut OutputHandle>> = Cell::new(None);
+    static LOG_FILE: Cell<Option<&'static mut OutputHandle>> = Cell::new(None);
+}
+
+pub(crate) fn reset() {
+    STANDARD_OUTPUT.with(|cell| cell.set(None));
+    LOG_FILE.with(|cell| cell.set(None));
+}
+
+fn with_stdout<T>(f: impl FnOnce(&mut OutputHandle) -> T) -> T {
+    STANDARD_OUTPUT.with(|out| {
+        let mut stdout = out.replace(None);
+        let res = f(stdout.as_mut().unwrap());
+        out.set(stdout);
+        res
+    })
+}
+
+fn with_log<T>(f: impl FnOnce(&mut OutputHandle) -> T) -> T {
+    LOG_FILE.with(|out| {
+        let mut log = out.replace(None);
+        let res = f(log.as_mut().unwrap());
+        out.set(log);
+        res
+    })
+}
+
+pub(crate) fn write_logs<B: ?Sized + AsBytes>(str: &B) {
+    with_log(|log| log.write_all(str.as_bytes())).unwrap();
```

Comment:
Would these `unwrap()s` potentially panic on an I/O error?

Reply:
Yes, I think (memory, might be wrong) bibtex may have just ignored errors here. I can do any of the following:
- Ignore errors, exactly match bibtex
- Bubble up errors to match other bibtex error handling (may cause other issues if that causes more writes, probably not what we want here)
- Keep this behavior - panic instead of silently dropping missed writes

Reply:
I don't think that panicking is a desirable behavior here. It might be possible to bubble up the errors, but I have a strong suspicion that implementing that would be a big pain and could potentially introduce a recursion issue as you suggest. So I think that the least-bad approach here is to "handle" any errors within the function. I wouldn't love using `eprintln!()` to report any error, but it would be better than nothing, I think.

Reply:
I'm mostly worried with `eprintln` that it would just panic as well - if the log file fails it might work to report that on stdout, but if the stdout writing fails, it seems a bad bet that stdout writing will succeed. I know tectonic might 'fake' stdout through OutputHandle, but I think it's a fair assumption for a program to make that stdout is stdout.